File: src/agenthandler.cpp
#include "agenthandler.hpp"

// Constructor for the AgentHandler class
AgentHandler::AgentHandler(boost::asio::streambuf& readBuffer,
                           boost::asio::streambuf& writeBuffer,
                           const std::shared_ptr<Config>& config,
                           const std::shared_ptr<Log>& log,
                           const TCPClient::pointer& client)
    : config_(config),            // Initialize configuration
      log_(log),                  // Initialize logging
      client_(client),            // Initialize TCP client
      readBuffer_(readBuffer),    // Reference to the read buffer
      writeBuffer_(writeBuffer),  // Reference to the write buffer
      request_(HTTP::create(config, log, readBuffer)) {
}  // Initialize HTTP request handler

// Destructor for the AgentHandler class
AgentHandler::~AgentHandler() {}

// Main handler function for processing requests
void AgentHandler::handle() {
  // Initialize encryption status
  BoolStr encryption{false, std::string("FAILED")};

  // Encrypt the request data from the read buffer
  encryption =
      aes256Encrypt(hexStreambufToStr(readBuffer_), config_->agent().token);

  if (encryption.ok) {
    // Log successful token validation
    log_->write("[AgentHandler handle] [Token Valid]", Log::Level::DEBUG);

    // Generate a new HTTP POST request string with the encrypted message
    std::string newReq(
        request_->genHttpPostReqString(encode64(encryption.message)));

    // Check if the request type is valid
    if (request_->detectType()) {
      // Log the HTTP request details
      log_->write("[AgentHandler handle] [Request] : " + request_->toString(),
                  Log::Level::DEBUG);

      // If the client socket is not open or the request type is HTTP or CONNECT
      if (!client_->socket().is_open() ||
          request_->httpType() == HTTP::HttpType::http ||
          request_->httpType() == HTTP::HttpType::connect) {
        // Connect the TCP client to the server
        boost::system::error_code ec;
        client_->doConnect(config_->agent().serverIp,
                           config_->agent().serverPort);

        // Check for connection errors
        if (ec) {
          log_->write(std::string("[AgentHandler handle] Connection error: ") +
                          ec.message(),
                      Log::Level::ERROR);
          return;
        }
      }

      // Copy the new request to the read buffer
      copyStringToStreambuf(newReq, readBuffer_);

      // Log the request to be sent to the server
      log_->write("[AgentHandler handle] [Request To Server] : \n" + newReq,
                  Log::Level::DEBUG);

      // Write the request to the client socket and initiate a read operation
      client_->doWrite(readBuffer_);
      client_->doRead();

      // Check if there is data available in the read buffer
      if (client_->readBuffer().size() > 0) {
        // If the HTTP request type is not CONNECT
        if (request_->httpType() != HTTP::HttpType::connect) {
          // Create an HTTP response handler
          HTTP::pointer response =
              HTTP::create(config_, log_, client_->readBuffer());

          // Parse the HTTP response
          if (response->parseHttpResp()) {
            // Log the response details
            log_->write(
                "[AgentHandler handle] [Response] : " + response->restoString(),
                Log::Level::DEBUG);

            // Decrypt the response body
            BoolStr decryption{false, std::string("FAILED")};
            decryption =
                aes256Decrypt(decode64(boost::lexical_cast<std::string>(
                                  response->parsedHttpResponse().body())),
                              config_->agent().token);

            // Check if decryption was successful
            if (decryption.ok) {
              // Copy the decrypted message to the write buffer
              copyStringToStreambuf(decryption.message, writeBuffer_);
            } else {
              // Log decryption failure and close the socket
              log_->write("[AgentHandler handle] [Decryption Failed] : [ " +
                              decryption.message + " ] ",
                          Log::Level::DEBUG);
              log_->write("[AgentHandler handle] [Decryption Failed] : " +
                              request_->toString(),
                          Log::Level::INFO);
              client_->socket().close();
            }
          } else {
            // Log if the response is not an HTTP response
            log_->write(
                "[AgentHandler handle] [NOT HTTP Response] [Response] : " +
                    streambufToString(client_->readBuffer()),
                Log::Level::DEBUG);
          }
        } else {
          // Log the response to a CONNECT request
          log_->write("[AgentHandler handle] [Response to connect] : \n" +
                          streambufToString(client_->readBuffer()),
                      Log::Level::DEBUG);

          // Move the response from the read buffer to the write buffer
          moveStreambuf(client_->readBuffer(), writeBuffer_);
        }
      } else {
        // Close the socket if no data is available
        client_->socket().close();
      }
    } else {
      // Log if the request is not a valid HTTP request
      log_->write("[AgentHandler handle] [NOT HTTP Request] [Request] : " +
                      streambufToString(readBuffer_),
                  Log::Level::DEBUG);
    }
  } else {
    // Log encryption failure and close the socket
    log_->write("[AgentHandler handle] [Encryption Failed] : [ " +
                    encryption.message + " ] ",
                Log::Level::DEBUG);
    log_->write(
        "[AgentHandler handle] [Encryption Failed] : " +
            client_->socket().remote_endpoint().address().to_string() + ":" +
            std::to_string(client_->socket().remote_endpoint().port()) + " ] ",
        Log::Level::INFO);
    client_->socket().close();
  }
}

File: src/agent_handler.rs
//! Agent-side request handling. Ports `agenthandler.cpp`.
//!
//! Takes the raw client request, wraps it (hex → AES → base64) inside a fake
//! outer HTTP POST, sends it to the nipo-server over the persistent `server`
//! connection, then decrypts the server's reply into bytes to return to the
//! client.

use tokio_rustls::TlsConnector;

use crate::client::{find_subslice, Conn};
use crate::config::Config;
use crate::crypto::{aes256_decrypt, aes256_encrypt, bytes_to_hex, encode64};
use crate::http::{host_from_url, sni_from_url, HttpType, ParsedHttp};
use crate::log::{Level, Log};

pub struct AgentOutcome {
    /// Bytes to write back to the downstream client.
    pub write_back: Vec<u8>,
    /// True when this was a CONNECT (tunnel should follow).
    pub connect: bool,
}

/// Process one client request through the agent. Returns `Ok(None)` to signal
/// the connection should be torn down (matches the C++ `socketShutdown` paths).
pub async fn handle(
    config: &Config,
    log: &Log,
    server: &mut Conn,
    connector: Option<&TlsConnector>,
    uuid: &str,
    src: &str,
    raw: &[u8],
) -> anyhow::Result<Option<AgentOutcome>> {
    let parsed = match ParsedHttp::detect(raw) {
        Some(p) => p,
        None => {
            log.write(
                &format!(
                    "[{uuid}] [AgentHandler handle] [NOT HTTP Request] [Request] : {}",
                    String::from_utf8_lossy(raw)
                ),
                Level::Debug,
            );
            return Ok(None);
        }
    };

    if !parsed.target.is_empty() {
        log.write(
            &format!("[{uuid}] [FORWARD] [SRC {src}] [DST {}]", parsed.target),
            Level::Trace,
        );
    }

    // Ensure the link to the nipo-server is up (reused across requests when
    // connectionReuse is on).
    if !server.is_open() {
        let server_ip = &config.agent.server_ip;
        let server_port = config.agent.server_port;
        let connected = if config.general.tls_enable {
            let sni = sni_from_url(&config.random_fake_url());
            let connector = connector
                .ok_or_else(|| anyhow::anyhow!("tls enabled but no connector"))?;
            server.connect_tls(server_ip, server_port, connector, &sni).await
        } else {
            server.connect_plain(server_ip, server_port).await
        };

        if let Err(e) = connected {
            log.write(
                &format!(
                    "[{uuid}] [CONNECT] [ERROR] [To Server] [SRC {src}] [DST {server_ip}:{server_port}] {e}"
                ),
                Level::Info,
            );
            return Ok(None);
        }
        log.write(
            &format!("[{uuid}] [CONNECT] [SRC {src}] [DST {server_ip}:{server_port}]"),
            Level::Info,
        );
    }

    // hex(raw) -> AES -> base64
    let encrypted = match aes256_encrypt(bytes_to_hex(raw).as_bytes(), config.general.token.as_bytes()) {
        Ok(c) => c,
        Err(e) => {
            log.write(
                &format!("[{uuid}] [AgentHandler handle] [Encryption Failed] : [ {e}] "),
                Level::Debug,
            );
            return Ok(None);
        }
    };
    let inner_request = encode64(&encrypted);

    let is_connect = parsed.http_type == HttpType::Connect;
    let action = if is_connect { "open" } else { "request" };
    let direct_tunnel_open = config.general.tunnel_enable && is_connect;
    let keep_alive = config.general.connection_reuse || direct_tunnel_open;

    let outer = format!(
        "{method} /{endpoint} HTTP/{ver}\r\n\
         Host: {host}\r\n\
         User-Agent: {ua}\r\n\
         Accept: */*\r\n\
         Content-Type: application/text\r\n\
         X-Nipo-Session: {uuid}\r\n\
         X-Nipo-Action: {action}\r\n\
         Content-Length: {clen}\r\n\
         Connection: {conn}\r\n\
         \r\n\
         {body}",
        method = config.random_method(),
        endpoint = config.random_end_point(),
        ver = config.agent.http_version,
        host = host_from_url(&config.random_fake_url()),
        ua = config.agent.user_agent,
        action = action,
        clen = inner_request.len(),
        conn = if keep_alive { "keep-alive" } else { "close" },
        body = inner_request,
    );

    if let Err(e) = server.write_all(outer.as_bytes()).await {
        log.write(
            &format!("[{uuid}] [AgentHandler handle] [write failed] {e}"),
            Level::Debug,
        );
        return Ok(None);
    }

    let response = match server.read_http_message().await {
        Ok(r) => r,
        Err(_) => {
            if !config.general.connection_reuse {
                return Ok(None);
            }
            return Ok(None);
        }
    };

    let body_pos = match find_subslice(&response, b"\r\n\r\n") {
        Some(p) => p + 4,
        None => {
            log.write(
                &format!(
                    "[{uuid}] [AgentHandler handle] [NOT HTTP Response] [Response] : {}",
                    String::from_utf8_lossy(&response)
                ),
                Level::Debug,
            );
            return Ok(None);
        }
    };

    let encrypted_body = &response[body_pos..];
    if encrypted_body.is_empty() {
        return Ok(None);
    }

    let decrypted = match aes256_decrypt(encrypted_body, config.general.token.as_bytes()) {
        Ok(d) => d,
        Err(e) => {
            log.write(
                &format!("[{uuid}] [AgentHandler handle] [Decryption Failed] : [ {e}] "),
                Level::Debug,
            );
            return Ok(None);
        }
    };

    log.write(
        &format!("[{uuid}] [AgentHandler handle] [Decryption Done]"),
        Level::Debug,
    );

    Ok(Some(AgentOutcome {
        write_back: decrypted,
        connect: is_connect,
    }))
}
